Fix #1342: Handle value classes for type parameters and don't unbox value classes returned as interface/supertype

ianbrandt opened the pull request:
Handle generic type parameters of value classes. Don't unbox value classes that are returned as their interface or supertype to avoid a `ClassCastException`. See added KDoc for details. Fixes mockk#1342.

ianbrandt commented on .gitignore:
```diff
@@ -6,6 +6,8 @@ build/
 
 .gradle/
 
+.kotlin/
```
Created by Kotlin Gradle Plugin during tests.

ianbrandt commented on test-modules/client-tests/src/jvmTest/kotlin/io/mockk/core/ValueClassSupportTest.kt:
```diff
@@ -20,8 +22,345 @@ class ValueClassSupportTest {
         assertEquals(JavaEnum.A, result)
     }
 
+    @Test
```
Added several test cases covering value class handling.

Raibaz commented:
@ianbrandt , can you please rebase on top of the master branch? I should have fixed the build failure.

ianbrandt commented:
@Raibaz, I finally got a chance to clean up the KDoc and inline comments for this change. I also rebased the changes onto the latest master (hence the force push to this branch). Let me know if you see anything else that needs fixing.

ianbrandt commented:
Done. I wasn't familiar with Spotless. I'll be sure to run that after any other changes going forward. Thanks!

ianbrandt commented:
> Looks like there are some test failures in older Kotlin versions and in instrumented tests. Can you PTAL?

I believe the client test failures with earlier versions of Kotlin were due to them being run with the latest Kotlinx Coroutines version, which is not that backwards compatible. I believe any `coEvery` in `:test-modules:client-tests` would've failed, there just wasn't any such coverage previously. I changed the build to use the latest compatible Kotlinx Coroutines version for each Kotlin version. The corresponding `coroutinesVersion` logic in 'test-modules/client-tests/build.gradle.kts' may be a point of maintenance going forward.

I believe the instrumented tests were failing because they require compatibility with API level 21, and I had used `java.lang.invoke.MethodHandle`, which only became available from API level 26. I changed to using `java.lang.reflect.Method` instead. All the `check` tests still pass. I'm not familiar with how to setup to run `connectedCheck` locally yet. I'll be watching for whether they pass if you approve the next workflow run.

P.S. I also rebased against current master and checked `spotlessApply`.

Raibaz commented:
Thanks a lot for all the follow ups, this looks very clean now! 🙏
